flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java

```diff
@@ -906,11 +907,33 @@ private void resetAvailableTransactionalIdsPool(Collection<String> transactional
 	// ----------------------------------- Utilities --------------------------
 
 	private void abortTransactions(Set<String> transactionalIds) {
-		for (String transactionalId : transactionalIds) {
-			try (FlinkKafkaProducer<byte[], byte[]> kafkaProducer =
-					initTransactionalProducer(transactionalId, false)) {
-				// it suffice to call initTransactions - this will abort any lingering transactions
-				kafkaProducer.initTransactions();
+		// shortcut for non-exactly-once producers
+		if (transactionalIds.isEmpty()) {
+			return;
+		}
+
+		ForkJoinPool forkJoinPool = null;
+		try {
+			// limit the number of connections to the number that is used during runtime
+			forkJoinPool = new ForkJoinPool(kafkaProducersPoolSize);
```

Comment:
Hmm, when we were discussing offline I understood that this will be global pool with static size. As it is now, we can be creating some substantial number (hundreds) of threads (albeit short lived). I'm not sure about this...

Crazy idea... what about making this pool `static` and with killing/shutting down idling threads?